src/utils/system_utils/src/system_utils.c

```diff
@@ -739,3 +740,75 @@ _Bool SystemUtils_IsFile(const char* path, int* err)
     }
     return is_file;
 }
+
+/**
+ * @brief For each dir_name in the baseDir, it calls the action function with the baseDir and dir_name.
+ * @param baseDir The base dir to list directories in.
+ * @param excludedDir A dir to exclude via exact match in addition to . and .. dirs. NULL means to exclude only . and .. dirs.
+ * @param perDirActionFunctor The functor to apply for each dir in the baseDir.
+ * @param err the error code (optional, can be NULL)
+ * @returns 0 if succeeded in calling the action func for every dir in baseDir (except . and .. dirs).
+ */
+int SystemUtils_ForEachDir(
+    const char* baseDir, const char* excludedDir, ADUC_SystemUtils_ForEachDirFunctor* perDirActionFunctor)
+{
+    int err_ret = -1;
+    struct dirent* dir_entry = NULL;
+    DIR* dir = NULL;
+
+    if (baseDir == NULL || (perDirActionFunctor == NULL) || (perDirActionFunctor->callbackFn == NULL))
+    {
+        goto done;
+    }
+
+    dir = opendir(baseDir);
+    if (dir == NULL)
+    {
+        err_ret = errno;
+        Log_Error("opendir '%s' failed: %d", baseDir, err_ret);
+        goto done;
+    }
+
+    do
+    {
+        errno = 0;
+        if ((dir_entry = readdir(dir)) != NULL)
+        {
+            if (strcmp(dir_entry->d_name, ".") == 0 || strcmp(dir_entry->d_name, "..") == 0)
+            {
+                continue;
+            }
+
+            if (excludedDir != NULL && strcmp(dir_entry->d_name, excludedDir) == 0)
+            {
+                continue;
+            }
+
+            perDirActionFunctor->callbackFn(perDirActionFunctor->context, baseDir, &(dir_entry->d_name)[0]);
+        }
+        else
+        {
+            if (errno != 0)
+            {
+                // error
+                err_ret = errno;
+                goto done;
+            }
+            else
```

Comment:
Do we need this 'else' block? You already goto done at line 794 anyway.

Reply:
removed but left some comment